### PR TITLE
s_server: correctly handle 2^14 byte long records

### PR DESCRIPTION
**Pull request by tomato42**
as the code uses BIO_gets, and it always null terminates the
strings it reads, when it reads a record 2^14 byte long, it actually
returns 2^14-1 bytes to the calling application

This makes the code sub-optimal (as every 2^14 record will need two
BIO_gets() calls) and makes it impossible to use -rev option to test
all plaintext lengths (like in openssl#15706)

##### Checklist

- [ ] tests are added or updated

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

File: apps/s_server.c [hunk 1]
```diff
@@ -3000,7 +3000,9 @@ static int www_body(int s, int stype, int prot, unsigned char *context)
     /* Set width for a select call if needed */
     width = s + 1;
 
-    p = buf = app_malloc(bufsize, "server www buffer");
+    /* as we use BIO_gets(), and it always null terminates data, we need
+     * to allocate 1 byte longer buffer to fit the full 2^14 byte record */
+    p = buf = app_malloc(bufsize + 1, "server www buffer");
```
Comment (mattcaswell) [inline apps/s_server.c:3005]: Wouldn't it be better to just change the definition of `bufsize` to be the *real* buffer size?

Comment (tomato42) [reply apps/s_server.c:3005]: not given the other places its used in, the www and rev is the exception not the rule in s_server

Comment (mattcaswell) [reply apps/s_server.c:3005]: ok

**Comment by tomato42** [comment]
changed the BIO_gets() calls more, the original code is actually incorrect, as BIO_gets() implicitly reads up to **size** -1 into the buffer as it leaves the 1 for the NUL-termination
In other words, BIO_gets() expects real size of *buf*, not max amount of data to read.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master branch. Thank you for your contribution.

This does not apply cleanly to 3.0 branch. I do not see this as something that needs to be backported there, but if you create a backport PR it can be merged.

**Pull request closed without merging**